### PR TITLE
dataloader fixex

### PR DESCRIPTION
**Pull request by UBCMark**

### CODE REVIEW

File: model_ptr.py [hunk 1]
```diff
@@ -130,46 +143,75 @@ def forward(self, features, captions, lengths, encoder_input, encoder_output):
...
         outputs = p_gen * outputs_regular + (1 - p_gen) * outputs_pointer
-        
-        outputs = pack(outputs, lengths, batch_first=True)[0] 
+
+        outputs = pack(outputs, lengths, batch_first=True)[0]
         return outputs
-    
-    def sample(self, features, states=None):
+
+    def sample(self, features, encoder_input, encoder_output, states=None):
```
Comment (UBCMark) [inline model_ptr.py]: This is done by @Wendy-Xiao . I modified model.ptr based on her version

File: data_loader.py [hunk 2]
```diff
@@ -52,8 +55,7 @@ def __init__(self, root, coco_annotation, vocab, MSCOCO_result, coco_detection_r
                     self.widths[key] = layouts['width']
                     self.heights[key] = layouts['height']
 
-            
-            train_visual_features = h5py.File('./data/train2014_visual_features.hdf5', 'r')
+            train_visual_features = h5py.File('/home/kwx/OBJ2TEXT-Improved/data/train2014_visual_features.hdf5', 'r')
```
Comment (TamakiSakura) [inline data_loader.py]: use relative path

File: data_loader.py [hunk 3]
```diff
@@ -191,17 +191,20 @@ def encode_location(bboxs, img_w, img_h):
...
-def get_loader(root, coco_annotation, vocab, 
-               MSCOCO_result, coco_detection_result, 
-               transform, batch_size, shuffle, num_workers, 
-               yolo=True, dummy_object=0):
+
+def get_loader(root, coco_annotation, vocab,
+               MSCOCO_result, coco_detection_result,
+               transform, batch_size, yolo,
+               shuffle, num_workers,
+               dummy_object=0):
```
Comment (TamakiSakura) [inline data_loader.py]: Explain the reason of changes in parameter orders - this affect all train*.py files.

File: train_ptr.py [hunk 4]
```diff
@@ -99,17 +99,17 @@ def main(args):
...
             # layout_encoder.zero_grad()
             # encoder.zero_grad()
-            
+
             # Modify This part for using visual features or not
-             
+
             # features = encoder(images)
-            layout_encoding, encoder_features = layout_encoder(label_seqs, location_seqs, layout_lengths)
+            layout_encoding, encoder_features = layout_encoder(label_seqs, location_seqs, visual_seqs, layout_lengths)
```
Comment (TamakiSakura) [inline train_ptr.py]: Not now

File: train_gtf.py [hunk 5]
```diff
@@ -133,18 +137,18 @@ def main(args):
                         help='path for saving trained models')
     parser.add_argument('--crop_size', type=int, default=224,
                         help='size for randomly cropping images')
-    parser.add_argument('--vocab_path', type=str, default='./data/vocab.pkl',
+    parser.add_argument('--vocab_path', type=str, default='/home/kwx/OBJ2TEXT-Improved/data/vocab.pkl',
```
Comment (TamakiSakura) [inline train_gtf.py]: Use relative path, absolute path is not allowed

File: train_gtf.py [hunk 6]
```diff
@@ -43,13 +43,13 @@ def main(args):
         vocab = pickle.load(f)
     Constants.set_constant(vocab)
 
+    yolo = True
     # Build data loader
-    data_loader = get_loader(args.image_dir, args.caption_path, vocab, 
+    data_loader = get_loader(args.image_dir, args.caption_path, vocab,
                              args.MSCOCO_result, args.coco_detection_result,
-                             transform, args.batch_size,
+                             transform, args.batch_size, yolo,
```
Comment (TamakiSakura) [inline train_gtf.py]: I don't see any benefit of changing this

**Review by TamakiSakura — CHANGES_REQUESTED** [review]
Please resolve all issues below

File: data_loader.py (hunk 2 above)
Comment (UBCMark) [reply data_loader.py]: resolved

File: train_gtf.py (hunk 6 above)
Comment (UBCMark) [reply train_gtf.py]: in each iteration, we check if `yolo == False` then we set `visual_seq == None`.

File: data_loader.py (hunk 3 above)
Comment (UBCMark) [reply data_loader.py]: See why I added a yolo flag in train*.py

File: train_gtf.py [hunk 7]
```diff
@@ -138,7 +142,7 @@ def main(args):
     parser.add_argument('--image_dir', type=str, default='./data/resized2014',
                         help='directory for resized images')
     parser.add_argument('--caption_path', type=str,
-                        default='./data/annotations/captions_train2014.json',
+                        default='.data/annotations/captions_train2014.json',
```
Comment (TamakiSakura) [inline train_gtf.py]: ？？？？

File: train_gtf.py (hunk 6 above)
Comment (TamakiSakura) [reply train_gtf.py]: OK but we could still do yolo=yolo, why change the position of the parameter?

**Review by TamakiSakura — CHANGES_REQUESTED** [review]
Please check the below

**Pull request merged**